kafkabp/consumer.go

```diff
@@ -0,0 +1,228 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *sarama.ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg ConsumerConfig
+	sc  *sarama.Config
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
+
+	closed          int64
+	consumeReturned int64
+	offset          int64
+
+	wg sync.WaitGroup
+}
+
+// Consumer defines the interface of a consumer struct.
+type Consumer interface {
+	io.Closer
+
+	Consume(ConsumeMessageFunc, ConsumeErrorFunc) error
+
+	// IsHealthy returns false after Consume returns.
+	IsHealthy() bool
+}
+
+// NewConsumer creates a new Kafka consumer.
```

Comment:
We should also explain the differences between the two consumers here. I know in this PR we only added one of them so there's no "differences" yet, but we still should add some explanation :)